[Refactor][UI/UX] Cleanup battle-info ui code

SirzBenjie opened the pull request:
## What are the changes the user will see?
N/A

## Why am I making these changes?
There are mumblings of wanting to change parts of the battle ui. I took a look at the ui file to try to get an understanding of what is going on there, and it's in quite poor shape.

## What are the changes from a developer perspective?
Created a new directroy, `src/ui/battle-info`.
Moved the old `battle-info.ts` file into it.
Converted `BattleInfo` into an abstract class, as it should never be instantiated itself.

I did not change any code logic. This was entirely just moving things around and cleaning up poor syntax to replace with equivalent code:

I didn't even clean up suspicious logic, such as this:
```ts
const baseVariant = !doubleShiny ? pokemon.getVariant(true) : pokemon.variant;
```
As that is not the point of this PR

- Made it so all methods now pass the `cognitiveComplexity` biome lint, by splitting up methods into smaller methods.

- Made it so that all of the phaser methods that return `this` for chaining actually use chaining. This shortens the code and makes it clearer what is going on.

- Moved methods into the `PlayerBattleInfo` and `EnemyBattleInfo` class into their own files.
  - Any method of `BattleInfo` class that had functionality surrounded by ``(if !playerInfo)`` (or not negated) had that logic moved to an overridden method in the respective class that calls the base class's method before doing that logic.

- Removed `updateEffectiveness` from `Pokemon` and only into `EnemyPokemon` class, as we never look at effectiveness for the player pokemon.

- Created a new function in `utils/common.ts` (I didn't know a better spot to put this) for getting the shiny descriptor for a variant.
- Modified all of the `GameObject` mocks in `test/testUtils/mocksContainer` to properly return the `this` parameter as required for certain methods. Also added a few mocked methods that were missing and necessary.

Previously, there was this nasty method contributing ~12 points of cognitive complexity:
```ts
doubleShiny || baseVariant ? `${baseVariant === 2 ? i18next.t("common:epicShiny") : baseVariant === 1 ? i18next.t("common:rareShiny") : i18next.t("common:commonShiny")}${doubleShiny ? `/${pokemon.fusionVariant === 2 ? i18next.t("common:epicShiny") : pokemon.fusionVariant === 1 ? i18next.t("common:rareShiny") : i18next.t("common:commonShiny")}` : ""}`
```

I updated `pokemon-info-container`, which also had the exact same pattern, to use this method. That's why that file was touched here as well.

Instead of that long ternary chain, I just made a short and sweet method that resolves to the i18n key for the variant.
## Screenshots/Videos
N/A

## How to test the changes?
N/A

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test`)
  - ~~[ ] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?~~
- ~~[ ] Have I provided screenshots/videos of the changes (if applicable)?~~
  - [x] Have I made sure that any UI change works for both UI themes (default and legacy)?

DayKev commented:
<details><summary>Beta 84a2ce9</summary>
<p>

![image](https://github.com/user-attachments/assets/86430864-1b10-422e-972b-57b4b2e67394)

</p>
</details> 
<details><summary>This PR ce0a166</summary>
<p>

![image](https://github.com/user-attachments/assets/2bdd6909-97cc-4c61-95e1-9d26d8c58f01)

</p>
</details>

Wlowscha commented on src/utils/common.ts:
```diff
@@ -566,3 +567,18 @@ export function animationFileName(move: Moves): string {
+  switch (variant) {
+    case 2:
+      return i18next.t("common:epicShiny");
+    case 1:
+      return i18next.t("common:shiny");
+    case 0:
+      return i18next.t("common:normal");
```
Also, note that the same logic for the hover description on the shiny icon is present in SummaryUiHandler as well. So the new function should also be used here. Since the logic is apparently the same in all three cases, maybe it would be best to have a common function that directly returns the hover text.

Wlowscha replied:
![shiny-rare-wrong](https://github.com/user-attachments/assets/bd811074-16f9-40db-89d2-53f7f63653e2)

Just to show what goes wrong

SirzBenjie replied:
> Since the logic is apparently the same in all three cases, maybe it would be best to have a common function that directly returns the hover text.

Agreed yeah. Another PR, though.

xsn34kzx commented on src/utils/common.ts:
```diff
@@ -576,3 +577,18 @@ export function animationFileName(move: Moves): string {
+ */
+export function getShinyDescriptor(variant: Variant): string {
+  switch (variant) {
+    case 2:
+      return i18next.t("common:epicShiny");
+    case 1:
+      return i18next.t("common:rareShiny");
+    case 0:
+      return i18next.t("common:commonShiny");
```
Is there a precedent/preference for descending or ascending order on `switch` blocks?

SirzBenjie replied:
I haven't seen a preference one way or the other in the codebase.

xsn34kzx commented on src/ui/fight-ui-handler.ts:
```diff
@@ -391,7 +391,7 @@ export default class FightUiHandler extends UiHandler implements InfoToggle {
     const opponents = (globalScene.getCurrentPhase() as CommandPhase).getPokemon().getOpponents();
     opponents.forEach(opponent => {
-      opponent.updateEffectiveness(undefined);
+      (opponent as EnemyPokemon).updateEffectiveness(undefined);
```
```suggestion
      (opponent as EnemyPokemon).updateEffectiveness();
```

xsn34kzx commented on src/ui/battle-info/enemy-battle-info.ts:
```diff
@@ -0,0 +1,203 @@
+  // #endregion
+
+  override get statOrder(): Stat[] {
+    return [Stat.HP, Stat.ATK, Stat.DEF, Stat.SPATK, Stat.SPDEF, Stat.ACC, Stat.EVA, Stat.SPD];
```
Would it break anything to make this array an immutable, static class field in this and the `PlayerBattleInfo` class?

SirzBenjie replied:
I wanted to make it an immutable static class field.
However, we can't actually make it static _and_ have this exist in the base battle info class. And all battle info classes need to have a stat order.

I played around with it and it was really cumbersome.
Plus, this achieves the same effect.

xsn34kzx replied:
Ah, makes sense then.

xsn34kzx reviewed (commented):
Super minor suggestions and questions since I've been out of the game for a minute.

emdeann commented on src/ui/fight-ui-handler.ts:
```diff
@@ -279,7 +279,7 @@ export default class FightUiHandler extends UiHandler implements InfoToggle {
 
       pokemon.getOpponents().forEach(opponent => {
-        opponent.updateEffectiveness(this.getEffectivenessText(pokemon, opponent, pokemonMove));
+        (opponent as EnemyPokemon).updateEffectiveness(this.getEffectivenessText(pokemon, opponent, pokemonMove));
```
Given that we're assuming here that `pokemon` is a `PlayerPokemon` and its opponents are `EnemyPokemon`, I would prefer this to look like `PlayerPokemon.getOpponents()` returning an array of `EnemyPokemon` instead of having to do this type assertion. but that is debatably at best in scope here

SirzBenjie replied:
This makes sense, but I feel as though it should be in another PR. This one is bloated as is.

DayKev commented:
Is there any way to manually generate a diff between the deleted `battle-info.ts` file and the new files? I want to just evaluate the differences and not comment on all the other nonsense that wasn't touched.

DayKev commented on src/ui/battle-info/battle-info.ts:
```diff
@@ -0,0 +1,655 @@
+  }
+
+  /** Helper method used by the constructor to create the tera and shiny icons next to the name */
+  private constructIcons() {
```
Why separate out these specifically?

SirzBenjie replied:
Good point! I should separate out more of them! That will make it a bit easier to split out the differences.

SirzBenjie replied:
Okay I split them out and this also helped remove almost all instances of the checks against `isPlayer`.
I couldn't really decide on a nice way to handle it for the `setTypes` method.

DayKev commented on src/ui/battle-info/battle-info.ts:
```diff
@@ -0,0 +1,655 @@
+    this.box = globalScene.add.sprite(0, 0, this.getTextureName()).setName("box").setOrigin(1, 0.5);
+    this.add(this.box);
+
+    this.nameText = addTextObject(player ? -115 : -124, player ? -15.2 : -11.2, "", TextStyle.BATTLE_INFO)
```
There's still a bunch of `player ? a : b`, shouldn't that also be split and moved into the subclasses?

DayKev commented on src/ui/battle-info/player-battle-info.ts:
```diff
@@ -0,0 +1,211 @@
+    this.lastLevelExp = pokemon.levelExp;
+
+    this.statValuesContainer.setPosition(8, 7);
+  }
```
```suggestion
  }

```

DayKev commented on src/ui/battle-info/player-battle-info.ts:
```diff
@@ -0,0 +1,211 @@
+
+  override get statOrder(): Stat[] {
+    return [Stat.ATK, Stat.DEF, Stat.SPATK, Stat.SPDEF, Stat.ACC, Stat.EVA, Stat.SPD];
+  }
```
```suggestion
  }

```

DayKev commented on src/ui/battle-info/battle-info.ts:
```diff
@@ -0,0 +1,655 @@
+    }
+  }
+
+  /** Update the pokemon name inside the container ,*/
```
```suggestion
  /** Update the pokemon name inside the container */
```

DayKev commented on src/ui/battle-info/battle-info.ts:
```diff
@@ -0,0 +1,655 @@
+      .on("pointerout", () => globalScene.ui.hideTooltip());
+  }
+
+  /** Called by {@linkcode initInfo} to initialize the shiny icon
```
```suggestion
  /**
   * Called by {@linkcode initInfo} to initialize the shiny icon
```

DayKev commented on src/ui/battle-info/battle-info.ts:
```diff
@@ -0,0 +1,655 @@
+
+  //#region Update methods and helpers
+
+  /** Update the status icon to match the pokemon's current status
```
```suggestion
  /**
   * Update the status icon to match the pokemon's current status
```

DayKev commented on src/ui/battle-info/player-battle-info.ts:
```diff
@@ -0,0 +1,211 @@
+    });
+  }
+
+  /** Updates the info on the info bar.
```
```suggestion
  /**
   * Updates the info on the info bar.
```

SirzBenjie commented:
> The HP number for the player UI doesn't fade in/out any more, instead appearing and disappearing instantly when C is pressed/released.

Thanks for pointing this out. The caught indicator and champion ribbon on enemy pokemon also had the same issue.
I fixed both by ensuring they are positioned below the stat info container.

I also addressed each of the comments from your review.
